monolith/src/monolith_mds.rs

```diff
@@ -0,0 +1,94 @@
+//! Monolith-31's default MDS permutation.
+//! With significant inspiration from https://extgit.iaik.tugraz.at/krypto/zkfriendlyhashzoo/
+
+use p3_field::PrimeField32;
+use p3_mds::util::apply_circulant;
+use p3_mds::MdsPermutation;
+use p3_mersenne_31::Mersenne31;
+use p3_symmetric::permutation::{ArrayPermutation, CryptographicPermutation};
+use sha3::digest::{ExtendableOutput, Update};
+use sha3::{Shake128, Shake128Reader};
+
+use crate::util::get_random_u32;
+
+#[derive(Clone)]
+pub struct MonolithMdsMatrixMersenne31<const NUM_ROUNDS: usize>;
```

Comment:
Hm this seems unusual in a way, normally we'd put any useful MDS impls in the `mds` crate, so they can be plugged into anything that needs MDS.

I don't quite get why these two particular MDS impls should be tied to Monolith; is it something where we want to match another impl for the sake of testing? If so, maybe it should be crate-private?

Reply:
Ah yes exactly, it's to match the reference implementation

Reply:
But isn't it useful to have this available? Because when we use a different matrix, then it's not _technically_ Monolith?

Reply:
Hm I think it's a bit of a gray area, like it was in the "spec" section of the paper, but it "suggested" a random Cauchy matrix and didn't mention details about randomness.

I guess there's no harm in keeping it since it's optional, we can revisit later when it becomes more clear whether this will become a standard or not.

Reply:
In general I am not sure whether we should allow for any MDS impl to be usable by any symmetric primitive that needs one. This opens an additional potential backdoor for attacks against the primitive using them. For instance some constructions like [Poseidon2](https://eprint.iacr.org/2023/323.pdf) or [Griffin](https://eprint.iacr.org/2022/403.pdf) don't even use an MDS matrix for their linear layer (only the inner matrix is MDS). Although this (is/may be) fine for their construction, it isn't necessarily for all, and this seems to pose a security concern for projects building on top of plonky3 with developers not having extended knowledge of this area.

In general, it seems that most recent constructions do fix the MDS layer out of caution, unlike previous designs like Rescue-Prime / old Poseidon, which were simply stating "One can use any MDS matrix".
As shown against the original specification of Poseidon [cf. [eprint/2020/188](https://eprint.iacr.org/2020/188.pdf)], some MDS matrices can actually be poor choices for the security of a symmetric primitive. This doesn't mean that we can't modify them of course, but then we're technically not respecting the specs anymore and happens what may 😁.

Reply:
On a different though related note, I was wondering what was the goal behind providing all those (recent) primitives (I saw mentions of Poseidon2/Tip5/Monolith in the repo). For educational/prototyping purposes, I guess there's not much harm, but these primitives are still fairly new and especially under-studied. Without a proper cryptanalysis effort, I feel like we'd want at least a warning mention that those primitives may hide some unknown breach?

Reply:
Hm but AFAIK only the Poseidon paper mentions security reasons for their particular MDS generation process; it seems like Monolith, RPO, Tip5 only mention performance reasons for their MDS choices.

Maybe there's an argument for nothing-up-my-sleeve methods, if we think there might be future security concerns with sparse subsets of MDS transforms, but by that measure the interpolation-based MDS transforms in Plonky3 may be even better (no semi-arbitrary seeds like `Monolith`).

Reply:
It's more of a precaution argument. By fixing the MDS matrix, even if this is for perf. reason (we did the same for Anemoi with our lightweight MDS matrices), this reduces the attack surface. If this is left to implementers, than we *might* have an MDS matrix sampled from a weak subspace that yields practical issues for a given primitive.

Reply:
Good point, maybe these newer primitives shouldn't be deployed in the near term, depending on security requiremnets. I guess the motivation for Monolith was to see how its performance compared; if it ends up being faster maybe it can be considered by certain projects that have backup security measures.